lib/ansible/module_utils/aws/urls.py

```diff
@@ -0,0 +1,193 @@
+# Copyright: (c) 2018, Aaron Haaf <aabonh@gmail.com>
+# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
+
+import datetime
+import hashlib
+import hmac
+import operator
+
+from ansible.module_utils.urls import open_url
+from ansible.module_utils.ec2 import boto3_conn, get_aws_connection_info, HAS_BOTO3
+from ansible.module_utils.six.moves.urllib.parse import urlencode
+from boto3 import session
+
+def hexdigest(s):
+    """
+    Returns the sha256 hexdigest of a string after encoding.
+    """
+
+    return hashlib.sha256(s.encode('utf-8')).hexdigest()
+
+
+def format_querystring(params=None):
+    """
+    Returns properly url-encoded query string from the provided params dict.
+
+    It's specially sorted for cannonical requests
+    """
+
+    if not params:
+        return ""
+
+    # Query string values must be URL-encoded (space=%20). The parameters must be sorted by name.
+    return urlencode(sorted(params.items(), operator.itemgetter(0)))
+
+
+# Key derivation functions. See:
+# http://docs.aws.amazon.com/general/latest/gr/signature-v4-examples.html#signature-v4-examples-python
+def sign(key, msg):
+    '''
+    Return digest for key applied to msg
+    '''
+
+    return hmac.new(key, msg.encode('utf-8'), hashlib.sha256).digest()
+
+
+def get_signature_key(key, dateStamp, regionName, serviceName):
+    '''
+    Returns signature key for AWS resource
+    '''
+
+    kDate = sign(('AWS4' + key).encode('utf-8'), dateStamp)
+    kRegion = sign(kDate, regionName)
+    kService = sign(kRegion, serviceName)
+    kSigning = sign(kService, 'aws4_request')
+    return kSigning
+
+
+def get_aws_key_pair(module):
+    '''
+    Returns aws_access_key_id, aws_secret_access_key for a module.
+
+    Prefers the credentials from a profile over inline'd keys
+    '''
+
+    if not HAS_BOTO3:
+        module.fail_json("get_aws_key_pair requires boto3")
+
+    dummy, dummy, boto_params = get_aws_connection_info(module, boto3=True)
+
+    profile = boto_params.get('profile_name')
+
+    if profile:
```

Comment:
you need to do `get_credentials` even if `profile` is not set (IAM instance profiles mean that neither environment variables, module parameters nor profile contain the keys)

Reply:
I think this code should just be:

```
 dummy, dummy, boto_params = get_aws_connection_info(module, boto3=True)
 profile = boto_params.get('profile_name')
s = session.Session(profile_name=profile)(**boto_params)
credentials = s.get_credentials()
return credentials.access_key, credentials.secret_key, credentials.token
```

Reply:
You could also just return `credentials` rather than the three attributes

Reply:
I'm pretty sure you can just go 

```
s = session.Session(**boto_params)
```

Since profile_name would be in `boto_params` anyway or absent